### PR TITLE
[Perf] Add SM103 hierarchical fused HGRN scan

### PR DESCRIPTION
**Pull request by heiheiha798**
## Summary

This adds an exact-SM103 long-sequence path for fused recurrent HGRN. The existing serial forward and backward kernels expose only `ceil(D / BD) * N` programs and keep the full token recurrence in each program. For eligible `T >= 1024` workloads, the new path:

1. computes fp32 affine summaries for independent 128-token blocks;
2. scans those summaries once per sequence and dimension tile; and
3. replays each block independently to write outputs or gradients.

Forward and backward both preserve the existing `exp` helper, fp32 state accumulation, output/gradient dtypes, flattened int64 addressing, public signatures, and dense/packed-varlen semantics. Scratch is two fp32 tensors of shape `N x ceil(T / 128) x D`. Packed input uses the hierarchy only when `N * ceil(T / 128) <= T`, which bounds each scratch tensor to at most `T * D` elements; over-budget packed input uses the original serial path. All non-SM103 devices and `T < 1024` also retain the unchanged serial kernels.

This is intentionally a B300-specific optimization. It adds a reusable exact capability `(10, 3)` helper because the existing SM100 helper is major-only.

Fork engineering record: [issue #2](https://github.com/heiheiha798/flash-linear-attention/issues/2) and [draft PR #3](https://github.com/heiheiha798/flash-linear-attention/pull/3).

## Test plan

- Dependent tests selected by `scripts/find_dependent_tests.py`: Slurm 5077, 256 passed / 3 skipped across `tests/ops/test_hgrn.py`, `tests/models/test_modeling_hgrn.py`, `tests/models/test_hybrid_attention.py`, and `tests/layers/test_layer_cache_layer_idx.py`.
- Repaired exact head `9fa2988`, Slurm 5803, NVIDIA B300 SM103: `tests/ops/test_hgrn.py` passed (`15 passed`).
- Full frozen gate after the implementation-review fix, Slurm 5082, NVIDIA B300 SM103:

  ```bash
  python -m benchmarks.ops.verify --op fused_recurrent_hgrn \
    --base e47d5d20aeb5989b58a3738b872e7c288a9fb75f \
    --modes fwd fwdbwd
  ```

  Result: 13/13 passed. This covers dense and packed-varlen forward/backward, partial and non-power-of-two dimensions, fp32/fp16, initial/final state, final-state gradients, `dx`, `dg`, and NaN-poisoned allocations.

- Deterministic many-short packed regression, Slurm 5082: public bf16 forward/backward at `T=N=8192, D=2048` selected the serial fallback, matched the one-token recurrence samples, and peaked at 0.719 GiB.
- Supplemental fixed bf16 partial-block and uneven-packed stress: passed in Slurm 5068 against the existing naive reference.
- Current-base synthetic merge at repaired head `9fa2988`: `pre-commit run --all-files` passed all 11 hooks, including ruff, autopep8, header-independent metadata checks, and the forbidden block-pointer check.
- Exact repaired head `9fa2988`: `python scripts/check_header.py --check`, the no-OJ gate, `git diff --check`, and in-memory Python compilation of all four changed Python files passed.
- The repair adds dispatch coverage in `tests/ops/test_hgrn.py` and exports the additive `IS_NVIDIA_SM103` helper through `fla.utils`. Public HGRN operator signatures, naive/reference code, tolerances, and numeric flags remain unchanged. The existing matching HGRN tests continue to exercise dense and packed-varlen public behavior.

Slurm 5081 is retained as an invalid pre-validation attempt: the ignored regression harness raised a Python `NameError` before invoking the operator, so none of that job's data is claimed. The mechanical harness correction and unchanged validation matrix completed in Slurm 5082.

## Benchmark / NCU

Hardware/software: NVIDIA B300 SXM6 AC (SM103), driver 580.126.09, CUDA 13.0 / nvcc 13.0.88, PyTorch 2.13.0+cu130, Triton 3.7.1. Baseline and candidate used separately pinned source trees and caches in the same allocations with fixed inputs, warmups, and repetitions.

The full registered `SHAPE_CONFIGS` matrix and long-context controls were measured in two alternating 25-sample rounds in Slurm 5068. Per-shape median/mean/std/min/p10/p90 are retained in the ignored raw results at `profile/fused_recurrent_hgrn-opt/results/{base,candidate}-5068-{1,2}.json`.

| Matrix | Forward geomean | Forward+backward geomean | Worst target speedup |
| --- | ---: | ---: | ---: |
| registered `SHAPE_CONFIGS` | 5.196x | 4.407x | 1.72x / 1.60x |
| long-context controls | 20.417x | 16.586x | 6.51x / 5.32x |

Long-context controls were B1/T8192/D2048, B2/T4096/D2048, B4/T2048/D2048, and B1/T32768/D512. Every long target improved. The unchanged short fallback was checked separately in Slurm 5072 with four alternating 1000-sample rounds at B1/T32/D2048: 1.004x forward and 1.145x forward+backward.

The implementation-review guard preserves the measured uneven packed N=4/T8192/D2048 bucket. Slurm 5082 used independently pinned source trees, five warmups, and 50 samples per source:

| Mode | Baseline median | Candidate median | Speedup |
| --- | ---: | ---: | ---: |
| forward | 2.085 ms | 0.168 ms | 12.385x |
| forward+backward | 4.854 ms | 0.521 ms | 9.321x |

The repository-native training command was run from independently asserted source roots in Slurm 5079 after rejecting an earlier editable-install-contaminated attempt:

```bash
python benchmarks/benchmark_training_throughput.py \
  --name hgrn --num_hidden_layers 1 --warmup_steps 8 --steps 20 \
  --batch_size 1 --seq_len 8192
python benchmarks/benchmark_training_throughput.py \
  --name hgrn --num_hidden_layers 1 --warmup_steps 8 --steps 20 \
  --batch_size 2 --seq_len 4096 --varlen
```

| Workload | Baseline tokens/s | Candidate tokens/s | Speedup | Baseline peak | Candidate peak |
| --- | ---: | ---: | ---: | ---: | ---: |
| dense B1/T8192/L1 | 529,229 | 1,176,641 | 2.223x | 4.20 GiB | 4.20 GiB |
| packed-varlen B2/T4096/L1 | 1,046,162 | 1,154,317 | 1.103x | 4.20 GiB | 4.20 GiB |

A separate fixed-input same-allocation harness in Slurm 5071 used 3 warmups and 12 samples per source and retained median/mean/std/min/p10/p90 distributions under `profile/fused_recurrent_hgrn-opt/model-results/`. Median speedups were 2.340x dense L1 B1/T8192, 1.078x packed L1 total-T8192, 1.954x dense L4 B2/T4096, 1.117x packed L4 B2/T4096, and 1.435x for the unmodified default 24-layer B1/T2048 config. No workload regressed; peak allocation was unchanged in every pair. Those samples preceded only a mechanical import-spacing correction and the packed over-budget fallback; Slurm 5073, 5079, and 5082 validate the resulting source and still-eligible buckets.

Nsight Compute 2025.3.1 full and source evidence was collected one target per allocation with three warmups and one collected launch. The external NCU report helper was unavailable, so the repository-local minimal workflow was used. Raw `.ncu-rep` files and exports remain ignored under `profile/fused_recurrent_hgrn-opt/ncu-reports/job-{5074,5075,5076}-*`; they are not part of this PR. No standalone checksum artifacts are retained.

| Workload/path | Baseline stage sum | Candidate stage sum | Speedup | Programs | Occupancy | Eligible warps | Max regs/spills |
| --- | ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| dense fwd B1/T8192/D2048 | 4.195 ms | 0.159 ms | 26.38x | 64 -> 2064 | 3.1% -> 35.1% | 0.022 -> 0.266 | 23 / 0 |
| dense bwd B1/T8192/D2048 | 4.364 ms | 0.174 ms | 25.06x | 64 -> 2064 | 1.6% -> 37.8% | 0.033 -> 0.352 | 29 / 0 |
| packed fwd total-T8192/D2048 | 2.214 ms | 0.158 ms | 14.00x | 256 -> 8256 | 3.7% -> 36.9% | 0.024 -> 0.287 | 24 / 0 |
| packed bwd total-T8192/D2048 | 2.451 ms | 0.172 ms | 14.27x | 256 -> 8256 | 3.8% -> 39.0% | 0.036 -> 0.430 | 24 / 0 |

For eligible shapes, the hierarchy trades two extra launches and additional scratch DRAM/L2 traffic for 32.25x more summary/replay programs. Source counters show the serial hotspot in the full-length recurrence and the candidate hotspot in independent 128-token block loops. Occupancy and eligible warps rise without spills, branch divergence, skipped work, or changed math. The T32 and over-budget packed controls resolve to the original serial kernels.

Relevant jobs: baseline gate 5061; profile 5064; isolated A/B 5068; model A/B 5071 and 5079; short control 5072; final pre-review gate 5073; NCU 5074-5076; dependent tests 5077; invalid review harness 5081; accepted review fix 5082.

## Risks and rollback

- The long-SM103 path adds two launches and fp32 block scratch per direction; the measured `T >= 1024` and packed scratch-budget guards prevent this overhead on short or many-short packed sequences.
- Eligible workloads increase scratch and global traffic, but measured model allocation did not change and NCU confirms that exposed parallelism dominates the traffic cost.
- Other NVIDIA capabilities, non-NVIDIA backends, and over-budget packed layouts use the original kernels unchanged.
- Rollback is the removal of the exact-SM103 dispatch/helper and the six hierarchical kernels; no saved-state or public contract migration is needed.

## Breaking changes

None.

## Checklist

- [x] I have read [CONTRIBUTING.md](https://github.com/fla-org/flash-linear-attention/blob/main/CONTRIBUTING.md) and follow its conventions (code style, docstrings, commit prefixes).
- [x] I have read [AGENTS.md](https://github.com/fla-org/flash-linear-attention/blob/main/AGENTS.md) and, where my change matches its scope, the relevant skill under [.agents/skills](https://github.com/fla-org/flash-linear-attention/tree/main/.agents/skills).
- [x] This is not a minor/cosmetic-only PR (typo, formatting, style-only tweaks).
- [x] Dependent tests pass locally or in CI; new behavior is covered by tests where applicable.
- [x] Kernel changes include same-hardware before/after benchmark numbers (dense + varlen where applicable).

### CODE REVIEW

**Comment by zhiyuan1i** [comment]
Not mergeable — closing.

There is no SM103-specific code here. All six kernels are plain Triton over regular CUDA-core resources (`tl.load`/`tl.store`/`exp`); SM90, SM100 and SM103 are identical for this kind of code, and nothing touches tcgen05 or TMA. The `get_device_capability() == (10, 3)` gate has no technical basis — it only records which machine the benchmarks ran on. It also makes the new path execute zero times in upstream CI with zero runnable tests, wasting reviewer time on code nobody can verify.

FLA's philosophy is portable, reviewable kernels that every supported platform can test. A path branded with the newest architecture name, gated for no reason and untestable by any maintainer, breaks that.
